src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -539,28 +519,57 @@ async def determine_setup_kwargs(
     return await Get(SetupKwargs, SetupKwargsRequest, setup_kwargs_request(target))  # type: ignore[abstract]
 
 
+@dataclass(frozen=True)
+class GenerateSetupPyRequest:
+    exported_target: ExportedTarget
+    sources: SetupPySources
+
+
+@dataclass(frozen=True)
+class GeneratedSetupPy:
+    digest: Digest
+
+
 @rule
 async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
-    exported_target = request.exported_target
-    exported_addr = exported_target.target.address
+    sources = await Get(SetupPySources, SetupPyChrootRequest, request)
+    snapshot = await Get(Snapshot, Digest, sources.digest)
+    # TODO: Add a generate_setup field to the python_distribution target type, so that
+    #  we're not guessing user intent (and not relying on the build being specifically
+    #  setup.py-based, it could be setup.cfg).
+    if "setup.py" in snapshot.files:
+        chroot_digest = sources.digest
+    else:
+        logger.info("No setup.py found, generating one...")
+        generated_setup_py = await Get(
+            GeneratedSetupPy, GenerateSetupPyRequest(request.exported_target, sources)
+        )
+        chroot_digest = await Get(Digest, MergeDigests((sources.digest, generated_setup_py.digest)))
+    return SetupPyChroot(chroot_digest)
 
-    owned_deps, transitive_targets = await MultiGet(
-        Get(OwnedDependencies, DependencyOwner(exported_target)),
-        Get(
-            TransitiveTargets,
-            TransitiveTargetsRequest([exported_target.target.address]),
-        ),
-    )
-    # files() targets aren't owned by a single exported target - they aren't code, so
-    # we allow them to be in multiple dists. This is helpful for, e.g., embedding
-    # a standard license file in a dist.
-    files_targets = (tgt for tgt in transitive_targets.closure if tgt.has_field(FilesSources))
-    targets = Targets(itertools.chain((od.target for od in owned_deps), files_targets))
 
-    sources, requirements = await MultiGet(
```

Comment:
requirements are now only needed when we generate setup kwargs, so this whole chain that gets them is now encapsulated there.